### PR TITLE
New Shadow module

### LINKED ISSUE #806: Adding a shadow to the image
### Please describe the problem (or idea)

> What did you expect to see that you didn't?
Add a shadow to the image

### Please show us where to look
http://sequencer.publiclab.org/examples/#steps=

### What's your PublicLab.org username?
niravasher

### Browser, version, and operating system
windows 10 with ubuntu subsystem
****

## Thank you!

Your help makes Public Lab better! We *deeply* appreciate your helping refine and improve this site.

To learn how to write really great issues, which increases the chances they'll be resolved, see:

https://publiclab.org/wiki/developers#Contributing+for+non-coders
@jywarren

### PR DESCRIPTION
**Pull request by niravasher**
Fixes #806   (<=== Replace `0000` with the Issue Number)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/is-reviewers` for help, in a comment below
* [x] Insert-step functionality is working correct as expected.
> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software
Please make sure to get at least two reviews before asking for merging the PR as that would make the PR more reliable on our part
Thanks!

### CODE REVIEW

**Comment by niravasher** [comment]
`@publiclab/is-reviewers` @HarshKhandeparkar @jywarren @VibhorCodecianGupta please review this. Made it with the help of canvas resize module.
Screenshot from 2020-01-15 00-46-39

**Comment by HarshKhandeparkar** [comment]
Did you copy the code of canvas resize or do you use canvas resize module inside this module?

**Comment by niravasher** [comment]
This portion is from canvas resize, the rest is similar to that of gradient module. But this cannot be a metamodule of the two because here gradient is present only till 20 pixels. @HarshKhandeparkar 
```
for (let i = 0; i < iMax && i < w; i++) {
        for (let j = 0; j < jMax && j < h; j++) {
          let x = i + 20, y = j + 20;
          pixelSetter(x, y, [pixels.get(i, j, 0), pixels.get(i, j, 1), pixels.get(i, j, 2), pixels.get(i, j, 3)], newPixels);
        }
      }
```

**Comment by HarshKhandeparkar** [comment]
Don't you think it will be better to use canvas resize inside of this module because if a change is required (to the canvas resize code), ever in the future, we won't have to change it at two places.

Also, duplicate code is not good. Also, the code will be easier to keep track of. And easier to document.

**Comment by niravasher** [comment]
So you want me to create a metamodule with some extra lines that make up for the gradient? @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Not exactly a meta module. Just import `require('../CanvasResize/...')`.

Have a look at the code of EdgeDetect. It imports blur module in this
manner.

**Comment by niravasher** [comment]
Ohh, looking into it.

**Comment by blurry-x-face** [comment]
> Not exactly a meta module. Just import `require('../CanvasResize/...')`. Have a look at the code of EdgeDetect. It imports blur module in this manner.

We use InternalSequencer in edge detect.

@niravasher you can import resize and any other module in the same way

**Comment by HarshKhandeparkar** [comment]
> We use InternalSequencer in edge detect.

We don't do that anymore. We just use the import the Module.js and use the draw function.

**Comment by blurry-x-face** [comment]
> We don't do that anymore. We just use the import the Module.js and use the draw function.

Yea just saw! I thought we were still using InternalSequencer :sweat_smile:

**Comment by niravasher** [comment]
In edge detect, all pixels are made blur by using Blur module. Here I am taking a small part of canvasResize and inside pixelManipulation adding extra lines of code for making gradients. I am not sure that can be done by importing canvasResize module. @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Hmm. Does canvas resize return a datauri and not pixels? If that is
the case then you can keep the code as it is.

**Comment by niravasher** [comment]
Datauri it is @HarshKhandeparkar

**Comment by HarshKhandeparkar** [comment]
Ohk. Then your code is good. :+1:

**Comment by niravasher** [comment]
Please merge the PR @jywarren

**Comment by blurry-x-face** [comment]
Hi! This looks good! Could you write a module test to protect this in future?

**Comment by blurry-x-face** [comment]
Does this works with Gifs?

File: src/modules/Shadow/info.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "shadow",
+  "description": "Creates a two sided shadow.",
+  "inputs": {},
+  "docs-link":"https://github.com/publiclab/image-sequencer/blob/main/docs/MODULES.md#shadow-module"
```
Comment (blurry-x-face) [inline src/modules/Shadow/info.json]: This doesn't exist, Could you write the docs also?

**Comment by blurry-x-face** [comment]
At the moment it just adds the shadow to the top and left can we do something to let the user choose where user wants the shadow like css box-shadow property, ref: [W3Schools](https://www.w3schools.com/cssref/css3_pr_box-shadow.asp), that would be really cool!!!

File: src/modules/Shadow/info.json (hunk 1 above)
Comment (niravasher) [reply src/modules/Shadow/info.json]: Ill open a new issue for this. Any newcomer can work on documentation part as a good first time issue. @blurry-x-face

**Comment by niravasher** [comment]
> At the moment it just adds the shadow to the top and left can we do something to let the user choose where user wants the shadow like css box-shadow property, ref: [W3Schools](https://www.w3schools.com/cssref/css3_pr_box-shadow.asp), that would be really cool!!!

Yes this can be done @blurry-x-face will do this tmrw.

**Comment by niravasher** [comment]
I can now add two values for the shadow as per w3schools.
Screenshot from 2020-01-24 01-35-24

Still the module is not perfect and I am working on perfecting the area where both gradients meet. @blurry-x-face

**Comment by niravasher** [comment]
Module is ready with two inputs and proper intersection of gradients between the two shadows. Shadows will be created just as w3schools @HarshKhandeparkar @blurry-x-face

**Comment by niravasher** [comment]
@jywarren @HarshKhandeparkar @blurry-x-face

**Comment by blurry-x-face** [comment]
Awesome work @niravasher :tada: , could give a Gif to show its working?

**Comment by blurry-x-face** [comment]
Also, why is the test for shadow module failing?
https://travis-ci.org/publiclab/image-sequencer/jobs/641393190#L1407

File: test/core/modules/shadow.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
...BOzgWnntb4kQkrik0y+OTHNnrUQ0Ws20MmMkrTdaDRLITIwxQSohgjFM7yZ6EWogkIAQRlhBOUGitQCSzJ0WAhpIDIrzssF55dg75ghtofNHZjWIjiVQkApOmvjt3rhY4eP5kuPVxPHmmub3uNCOaqWNZK6B727ZfvVrRhZ+84BOC5SoMmPsrPvOuid2YQk1EdOOl4314ZHxcMc4b1z/yQ95/5Pv8d4P3+Prr6+5fXzJ/fnAkSOiERJqMHJgEnsAA3nQunAH1wZjh20lT2fytMBIch+ojKuwTbSgZDIHdDNfddpF4L5jnXE18IHCjBLJxDY6yy6Gg6++/oZf/vxfOb25g5FgsI0rwcYGY5CRBAZFQBnzjo0okBHCmCfGhAJjRNBUGJMUDQgFtpGEQsgmIkBGgtYaTyLgYmp88sELri+O7AIisKHKzIeZJrOPjXF/zzw3pvbIjz74iO//8D/y+U9+yv/63/6ef/zlryALCbKgAJUgZrLANmVA4OL/JdQCBEIogiBwCFdRY8djhyqkIEJgYychkJJw0tSIMoexMRkmissGB+BFiI+PM9+LM1fLW6b7V+S3XzLdv+LAxjY3Hj9+Hxva1TOmLL66f8O5X7KrERG0PqFoEB0DBkLC0XCbUINWok2BKE774NX9mc+/9xFMR7ZhHAeYgz51xjhTudBvjuxbkusbYjxD8zPcRFwciYsju+6IHkzHA0qgTpwfB6MK845BgG0wYGFDZZEyscPYE4YJRNUgZOiACtjAK67CuYMTubEtxevffMnv/u9f8fDr3+E3D/S9aGlqFH2aGNHwYSKe3VBX19wvg1/+5nf8y9//Cy+//JbxuBFMEEFEwzakyCFc4oklhpNCBAERZCW2QWAFuQ8UgRFEUJVEa/TWSaBUVBURYEAtmKbOUNK6aE1EF6JTghbBk7KxwWkocBkroKBcBKbJKHjHOAe1LewqXIkQNTY0NuSkCybDi9z58xfP+V9+9Dk/+egFh31l/PrXbPNzpo9+QDsULkGfCII+H8hmFMISCJgnpg8+pK8/4Pz7X8Ptay4PE2PZyPMOBjVRWdTYwUlEkKcHtpdfcnh2w3z9nK025MY8z1wcOq2ZvLuF5Y7YTuxvbuHhnhgDpVGBs3BBk5BEVRGt0VonxyARYWGbGhumITWiN9SEbLBBwi4kES3IsTMkPEH/5D0uf/AJ8cF7
```
Comment (blurry-x-face) [inline test/core/modules/shadow.js]: Change the options input to something other than default and update the benchmark accordingly!

Comment (niravasher) [reply test/core/modules/shadow.js]: From where should I get the benchmark? There are lot of websites available to convert img to datauri. So I get different datauri's. How do I match it with my output image?

Comment (blurry-x-face) [reply test/core/modules/shadow.js]: Just use your module to generate the image in IS then open the image in a new tab or you can use the sequencer object to get the data uri, sequencer.steps holds the array of all the steps you just go through it you'll find it easily! Probably in `sequencer.steps[1].output.src`

Comment (niravasher) [reply test/core/modules/shadow.js]: Done @blurry-x-face 
Screenshot from 2020-01-26 11-40-30

**Comment by niravasher** [comment]
ezgif com-gif-maker
@blurry-x-face

**Comment by niravasher** [comment]
Yes this does work with GIF's too.
image_1
@HarshKhandeparkar

**Comment by niravasher** [comment]
Anything else I need to do here? @HarshKhandeparkar ? If not please merge this.

**Comment by niravasher** [comment]
Pls merge this @jywarren @HarshKhandeparkar

File: src/modules/Shadow/Module.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        }
+        for (let i = 0; i < iMax && i < w; i++) {
+          for (let j = 0; j < jMax && j < h; j++) {
+            let x = i + Math.abs(options.X_value), y = j + Math.abs(options.Y_value);
+            pixelSetter(x, y, [pixels.get(i, j, 0), pixels.get(i, j, 1), pixels.get(i, j, 2), pixels.get(i, j, 3)], newPixels);
+          }
+        }
+      }
+
+      else if (options.X_value >= 0 && options.Y_value >= 0) {
```
Comment (HarshKhandeparkar) [inline src/modules/Shadow/Module.js:50]: Looks like there is a lot of repetition here, can that be fixed?

Comment (jywarren) [reply src/modules/Shadow/Module.js:50]: Do we feel this is blocking, or should we just open a new issue to address it? Thank you all! We could include this in `v3.7.0`!

Comment (HarshKhandeparkar) [reply src/modules/Shadow/Module.js:50]: Hmm, this review is quite old. We didn't get any reply so ig we can dismiss it :)

**Review by HarshKhandeparkar — APPROVED** [review]

**Review by jywarren — APPROVED** [review]

**Comment by jywarren** [comment]
Awesome!!!! Thanks so much!!!

**Pull request merged**